### PR TITLE
Audience membership core amplitude

### PR DESCRIPTION
**Pull request by joe-ayoub-segment**
Summary

  This PR introduces a centralized audienceMembership resolution mechanism in actions-core and migrates 1 audience destinations to consume it. 

Instead of each destination independently parsing Engage traits/properties or RETL sync-mode event names to determine add vs. remove, the Core framework now computes a typed boolean | undefined value and injects it into every action's ExecuteInput. All destination changes are gated behind per-destination feature flags with full legacy fallback.

  ---
  Core Changes (packages/core)

  New: audience-membership.ts

  Centralizes the logic for resolving audience membership from raw event data:

  - resolveAudienceMembership(rawData, syncMode) — top-level resolver; tries Engage first, then RETL
  - engageAudienceMembership(rawData) — reads context.personas.computation_class + computation_key from identify/track events emitted by Engage
  - retlAudienceMembership(rawData, syncMode) — maps RETL event names (new, updated, deleted) against the configured syncMode (add, update, upsert, mirror, delete)
  - Returns true (add), false (remove), or undefined (non-audience event / indeterminate)

  New: flags.ts
  Adds the following flags to Core. Will be removed once the associated destinations are migrated. 
  FLAGS.ACTIONS_GOOGLE_EC_AUDIENCE_MEMBERSHIP
  FLAGS.ACTIONS_BRAZE_COHORTS_AUDIENCE_MEMBERSHIP
  FLAGS.ACTIONS_LINKEDIN_AUDIENCES_AUDIENCE_MEMBERSHIP

  Updated: destination-kit/types.ts

  - Exports new AudienceMembership = boolean | undefined type
  - Adds optional audienceMembership?: AudienceMembershipType field to ExecuteInput — typed as AudienceMembership for perform and AudienceMembership[] for performBatch

  Updated: destination-kit/action.ts

  - Fixes a latent bug: syncMode values from mapping were not being validated through isSyncMode() before being passed to batch handlers

  Updated: index.ts

  Exports AudienceMembership, resolveAudienceMembership, and FLAGS for use by destinations.

  ---
  Destination Changes

**Amplitude Cohorts**

  Flag: ACTIONS_CORE_AUDIENCE_MEMBERSHIP (master flag only — amplitude has no per-destination flag; membership comes from Core unconditionally once the master flag is on)

  - syncAudience/fields.ts: Removed the engage_fields nested object (which contained segment_computation_class, traits_or_properties, segment_audience_key, and segment_external_audience_id). Replaced with a single
  top-level segment_external_audience_id field. This simplifies the mapping and removes the need for each payload to carry audience key / traits data.
  - syncAudience/functions.ts: send() now accepts audienceMemberships?: AudienceMembership[] and uses it to route payloads into addMap / deleteMap. Payloads with undefined membership are individually errored with
  PAYLOAD_VALIDATION_FAILED.
  - syncAudience/index.ts: Destructures audienceMembership from ExecuteInput and passes it to send().

  ---
  Unit Test Coverage

  - actions-core — New audience-membership.test.ts (362 lines) covering resolveAudienceMembership, engageAudienceMembership, and retlAudienceMembership across all syncModes, event types, and edge cases
  - Amplitude Cohorts — Rewrote syncAudience/__tests__/index.test.ts; preserved legacy behavior in new snapshot legacy-to-be-removed.test.ts.snap
  - Facebook Custom Audiences — Replaced monolithic index.test.ts + fb-operations.test.ts with purpose-built suites: mirror.test.ts (961 lines), upsert.test.ts (408 lines), delete.test.ts (537 lines),
  legacy.test.ts (295 lines), functions.test.ts (342 lines), audience-creation.test.ts (135 lines), canary.test.ts (175 lines)

Tested in Staging with Facebook Custom Audiences. 
image

### CODE REVIEW

**Review by AnkitSegment — APPROVED** [review]

**Review by abhandage — COMMENTED** [review]
## Code Review

Thanks for this — centralizing audience membership resolution in Core is the right architectural move. Good test coverage on `audience-membership.ts` and the `isSyncMode()` fix is a clean catch. A few issues need resolution before merging.

---

### Critical: Feature flag is tested but never gated

Tests pass `features: { 'actions-core-audience-membership': true }` throughout, but `action.ts` calls `resolveAudienceMembership()` **unconditionally** — neither the core action nor the amplitude destination ever reads `bundle.features` to gate on this flag. That means:

1. The flag value in tests is **ignored** — it documents intent that was never implemented (or was implemented and then reverted without cleaning up the tests).
2. `audienceMembership` is now computed on **every event across all ~213 destinations**, not just audience-aware ones.

If the intent is unconditional computation (reasonable, since `audienceMembership` is optional and only affects consumers that read it), remove the feature flag references from the tests so future readers aren't confused. If gating was intended, add the gate in `action.ts` before calling `resolveAudienceMembership()`.

---

### Critical: Breaking behavioral change for non-audience events in `perform()`

In `syncAudience/index.ts`, `[audienceMembership]` is passed where `audienceMembership` may be `undefined`. This passes the `Array.isArray` guard in `send()`, then hits the `membership === undefined` branch, which calls `handleError` with `isBatch = false` — throwing `PayloadValidationError`. The old code silently sent these as `REMOVE` operations (falsy `isAudienceMember` branch). That was arguably wrong, but this is a silent behavioral change with different retry semantics. If intentional, it should be called out explicitly.

---

### Medium: Dead exports in `flags.ts`

The three flags (`ACTIONS_GOOGLE_EC_AUDIENCE_MEMBERSHIP`, `ACTIONS_BRAZE_COHORTS_AUDIENCE_MEMBERSHIP`, `ACTIONS_LINKEDIN_AUDIENCES_AUDIENCE_MEMBERSHIP`) are exported from Core but never referenced — those destinations were reverted from this PR. Dead symbols in the public Core API are confusing. Either remove them now or add a comment explaining they are placeholders for follow-up PRs.

---

### Medium: Error message inconsistency

`functions.ts` line 25: `'Audience membership details missing'`
`functions.ts` line 48: `"Audience Membership Details missing"`

Same failure, different capitalization and wording.

---

### Positive

- `isSyncMode()` guard now applied consistently in both `execute()` and `executeBatch()` — latent bug fix is clean.
- Removing `engage_fields` + `traits_or_properties` from the payload schema is a good simplification.
- `testAuthentication` using `owner_email` is correct.
- `audience-membership.ts` is well-factored with solid independent test coverage.

**Comment by joe-ayoub-segment** [comment]
note - Amplitude Cohorts is in Private Beta and is still being tested with Amplitude team.

File: packages/destination-actions/src/destinations/amplitude-cohorts/index.ts [hunk 1]
```diff
@@ -59,10 +59,11 @@ const destination: AudienceDestinationDefinition<Settings, AudienceSettings> = {
...
     testAuthentication: (request, { settings }) => {
       const { 
-        endpoint 
+        endpoint,
+        owner_email
       } = settings
       const baseUrl = getEndpointByRegion('usersearch', endpoint)
-      return request(`${baseUrl}?user=testUser@example.com`)
+      return request(`${baseUrl}?user=${owner_email}`)
     }
```
[the remark this answers is not in the document]
Comment (joe-ayoub-segment) [reply packages/destination-actions/src/destinations/amplitude-cohorts/index.ts:67]: Amplitude Cohorts is private beta with no customer usage. 
Fix will be handled in a separate PR. Documented here for follow up: 
https://twilio-engineering.atlassian.net/browse/STRATCONN-6409

File: packages/core/src/flags.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+export const FLAGS = {
+    ACTIONS_GOOGLE_EC_AUDIENCE_MEMBERSHIP: 'actions-google-ec-audience-membership',
+    ACTIONS_BRAZE_COHORTS_AUDIENCE_MEMBERSHIP: 'actions-braze-cohorts-audience-membership',
+    ACTIONS_LINKEDIN_AUDIENCES_AUDIENCE_MEMBERSHIP: 'actions-linkedin-audiences-audience-membership'
+}
```
[the remark this answers is not in the document]
Comment (joe-ayoub-segment) [reply packages/core/src/flags.ts:6]: Not important enough to fix now.

File: packages/core/src/audience-membership.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  if (!computation_key) return undefined
+
+  let membershipValue: boolean | undefined
+
+  if (type === 'identify' && typeof traits?.[computation_key] === 'boolean') {
+    membershipValue = traits[computation_key]
+  } 
+  else if (type === 'track' && typeof properties?.[computation_key] === 'boolean') {
+    membershipValue = properties[computation_key]
+  }
```
[the remark this answers is not in the document]
Comment (joe-ayoub-segment) [reply packages/core/src/audience-membership.ts:61]: Prettier - low pri. Won't be fixed now.

File: packages/destination-actions/src/destinations/amplitude-cohorts/syncAudience/__tests__/getIds.test.ts [hunk 4]
```diff
@@ -55,42 +40,22 @@ describe('getIds function', () => {
...
           user_id: 'user1',
-          engage_fields: {
-            segment_computation_class: 'audience',
-            traits_or_properties: {},
-            segment_audience_key: 'test_audience',
-            segment_external_audience_id: 'cohort123'
-          },
+           segment_external_audience_id: 'cohort123',
           batch_size: 100
         },
```
[the remark this answers is not in the document]
Comment (joe-ayoub-segment) [reply packages/destination-actions/src/destinations/amplitude-cohorts/syncAudience/__tests__/getIds.test.ts:45]: Leading space issue - low pri. Won't fix now.

File: packages/core/src/destination-kit/action.ts [hunk 5]
```diff
@@ -456,15 +460,18 @@ export class Action<Settings, Payload extends JSONLikeObject, AudienceSettings =
     }
 
     if (this.definition.performBatch) {
-      const syncMode = this.definition.syncMode ? bundle.mapping?.['__segment_internal_sync_mode'] : undefined
+      const syncModeVal = this.definition.syncMode ? bundle.mapping?.['__segment_internal_sync_mode'] : undefined
+      const syncMode = isSyncMode(syncModeVal) ? syncModeVal : undefined
       const matchingKey = bundle.mapping?.['__segment_internal_matching_key']
+      const audienceMembership = bundle.data.map((d) => resolveAudienceMembership(d, syncMode))
```
[the remark this answers is not in the document]
Comment (joe-ayoub-segment) [reply packages/core/src/destination-kit/action.ts:466]: Fix [here](https://github.com/segmentio/action-destinations/pull/3825/changes#diff-52379ff63dd9caa3cfb44166dd85abd5e38f9f385cb391e8fddacf921b01040bR478) will address this issue.

**Comment by joe-ayoub-segment** [comment]
> ## Code Review
> Thanks for this — centralizing audience membership resolution in Core is the right architectural move. Good test coverage on `audience-membership.ts` and the `isSyncMode()` fix is a clean catch. A few issues need resolution before merging.
> 
> ### Critical: Feature flag is tested but never gated
> Tests pass `features: { 'actions-core-audience-membership': true }` throughout, but `action.ts` calls `resolveAudienceMembership()` **unconditionally** — neither the core action nor the amplitude destination ever reads `bundle.features` to gate on this flag. That means:
> 
> 1. The flag value in tests is **ignored** — it documents intent that was never implemented (or was implemented and then reverted without cleaning up the tests).
> 2. `audienceMembership` is now computed on **every event across all ~213 destinations**, not just audience-aware ones.
> 
> If the intent is unconditional computation (reasonable, since `audienceMembership` is optional and only affects consumers that read it), remove the feature flag references from the tests so future readers aren't confused. If gating was intended, add the gate in `action.ts` before calling `resolveAudienceMembership()`.
> 
> ### Critical: Breaking behavioral change for non-audience events in `perform()`
> In `syncAudience/index.ts`, `[audienceMembership]` is passed where `audienceMembership` may be `undefined`. This passes the `Array.isArray` guard in `send()`, then hits the `membership === undefined` branch, which calls `handleError` with `isBatch = false` — throwing `PayloadValidationError`. The old code silently sent these as `REMOVE` operations (falsy `isAudienceMember` branch). That was arguably wrong, but this is a silent behavioral change with different retry semantics. If intentional, it should be called out explicitly.
> 
> ### Medium: Dead exports in `flags.ts`
> The three flags (`ACTIONS_GOOGLE_EC_AUDIENCE_MEMBERSHIP`, `ACTIONS_BRAZE_COHORTS_AUDIENCE_MEMBERSHIP`, `ACTIONS_LINKEDIN_AUDIENCES_AUDIENCE_MEMBERSHIP`) are exported from Core but never referenced — those destinations were reverted from this PR. Dead symbols in the public Core API are confusing. Either remove them now or add a comment explaining they are placeholders for follow-up PRs.
> 
> ### Medium: Error message inconsistency
> `functions.ts` line 25: `'Audience membership details missing'` `functions.ts` line 48: `"Audience Membership Details missing"`
> 
> Same failure, different capitalization and wording.
> 
> ### Positive
> * `isSyncMode()` guard now applied consistently in both `execute()` and `executeBatch()` — latent bug fix is clean.
> * Removing `engage_fields` + `traits_or_properties` from the payload schema is a good simplification.
> * `testAuthentication` using `owner_email` is correct.
> * `audience-membership.ts` is well-factored with solid independent test coverage.

Addressing each of the items from this deep review. 

**Most importantly the deep review didn't detect the actual issue which caused the incident.** 


**This issue: Critical: Feature flag is tested but never gated**
 - This flag was deemed not needed and should have been removed from the original tests in the PR. The flag only existed in tests and didn't change the outcome of the tests. This will be updated to the new PR https://github.com/segmentio/action-destinations/pull/3825 

**This issue: Critical: Breaking behavioral change for non-audience events in perform()** 
- Not an issue. The events should fail if there is even one item where audience membership is undefined. 
- This should happen for single or batches. 

**This issue: Medium: Dead exports in flags.ts**
- This was to done deliberately so only one Core change would be needed. Separate PRs added code in destinations which use the flags. 
- This isn't an issue. 

**This issue: Medium: Error message inconsistency**
- this is a style issue. Low priority.
